refactor: pull "RaiseEvent" function from the middleware closure

MikelAlejoBR opened the pull request:
The reason for pulling it from the closure is that we might want to use the "RaiseEvent" function in a place where multiple events must be raised, and therefore if it is embedded in the closure that would be difficult to do.

MikelAlejoBR committed:
refactor: pull "RaiseEvent" function from the middleware closure

The reason for pulling it from the closure is that we might want to use
the "RaiseEvent" function in a place where multiple events must be
raised, and therefore if it is embedded in the closure that would be
difficult to do.

lindgrenj6 reviewed (commented):
Instead of just yoinking the event function and leaving it in the `middleware` package what do you think about moving it to `service` and make it take a couple parameters.

Something like `service.RaiseEvent(eventType string, thing model.Event)`, then the middleware would just call that function the params parsed from the context (any overrides, the message, the event itself). This way we would have separation of context-aware sending and one-off sending. 

Unless you were thinking of just setting the overrides in the handlers? but that wouldn't work for the source pause event I guess where you'd need to send N application events and 1 source event.

MikelAlejoBR committed:
refactor: move "RaiseEvent" to the service package

lindgrenj6 commented on middleware/event_stream.go:
```diff
@@ -51,70 +31,8 @@ func RaiseEvent(next echo.HandlerFunc) echo.HandlerFunc {
-		}
-
-		return nil
+		return service.RaiseEvent(c, eventType, resource)
```
Maybe I should have thought about it more. I know we're going to need to pass the context for the headers or just have another helper method that takes the context as an argument.

I was more implying we should leave the `event_override` and other stuff based on the context here in the middleware. Basically all that `service.RaiseEvent()` would do is marshal the resource + post the message. 

---

Maybe we can change the `getRequestHeaders(c)` function to something like `ForwardableHeaders(c)` so it takes in the echo context and returns the slice of headres - that way the new RaiseEvent function would have a signature like:

```go
func RaiseEvent(eventType string, resource m.Resource, headers []kafka.Header) error

// called via
service.RaiseEvent("Source.pause", source, service.ForwardableHeaders(c))
```

The RaiseEvent function would just add the `event_type` header and produce the message.

---

Does that make sense?

MikelAlejoBR replied:
Absolutely. I implemented those changes. Thanks for the feedback!

lindgrenj6 commented on middleware/event_stream.go:
```diff
@@ -1,46 +1,26 @@
-			l.Log.Infof("skipping raise event per skip_raise set on context")
-			return nil
-		}
-
```
So this part needs to stay - I am surprised my tests didn't catch it.

MikelAlejoBR replied:
I think it is because of the following:

```go
	f := raiseMiddleware(func(c echo.Context) error {
		c.Set("skip_raise", struct{}{})
		return c.NoContent(http.StatusNoContent)
	})
```

It seems to be returning the `StatusNoContent` all the time, which causes the tests to pass no matter what.

lindgrenj6 commented on middleware/event_stream.go:
```diff
@@ -51,70 +31,10 @@ func RaiseEvent(next echo.HandlerFunc) echo.HandlerFunc {
-			l.Log.Infof("Using overridden event_type %v instead of %v", c.Get("event_override"), eventType)
-			eventType = event
-		}
-
```
This should stay too. I don't have a test around this, oops!

lindgrenj6 reviewed (commented):
Just 2 little pieces that need to be added back into the middleware - other than that looks good to me!

lindgrenj6 commented on middleware/event_stream.go:
```diff
@@ -1,27 +1,15 @@
-func RaiseEvent(next echo.HandlerFunc) echo.HandlerFunc {
+// RaiseEventMiddleware calls the "RaiseEvent" function once the previous handler has succeeded. It grabs the resource
+// and the event type from the context.
+func RaiseEventMiddleware(next echo.HandlerFunc) echo.HandlerFunc {
```
One last teeny thing - this can now just be `RaiseEvent` again.

MikelAlejoBR replied:
Renamed!

MikelAlejoBR committed:
refactor: rename "RaiseEvent" function

Since they are separated in different packages, they can be named the
same in both packages.

MikelAlejoBR committed:
fix: typo
